pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -733,24 +734,31 @@ boolean populateMessageSchema(MessageImpl msg, SendCallback callback) {
             completeCallbackAndReleaseSemaphore(msg.getUncompressedSize(), callback, e);
             return false;
         }
-        byte[] schemaVersion = schemaCache.get(msg.getSchemaHash());
-        if (schemaVersion != null) {
-            msgMetadataBuilder.setSchemaVersion(schemaVersion);
-            msg.setSchemaState(MessageImpl.SchemaState.Ready);
-        }
+
+        // do not use the return value to return the function call.
+        setMessageSchema(msg);
+
         return true;
     }
 
-    private boolean rePopulateMessageSchema(MessageImpl msg) {
+    private boolean setMessageSchema(MessageImpl msg) {
         byte[] schemaVersion = schemaCache.get(msg.getSchemaHash());
         if (schemaVersion == null) {
             return false;
         }
-        msg.getMessageBuilder().setSchemaVersion(schemaVersion);
+
+        if (schemaVersion != SchemaVersion.Empty.bytes()) {
+            msg.getMessageBuilder().setSchemaVersion(schemaVersion);
+        }
+
         msg.setSchemaState(MessageImpl.SchemaState.Ready);
         return true;
     }
 
+    private boolean rePopulateMessageSchema(MessageImpl msg) {
+        return setMessageSchema(msg);
+    }
```

Comment:
I think we don't need to add another method (`setMessageSchema`) that is completely the same with the existing method (`rePopuateMessageSchema`)